src/backend/controllers/marketController.js

```diff
@@ -0,0 +1,308 @@
+        res.status(200).json(sale);
+    }
+    catch (err) {
+        console.error(err.message);
+        res.status(500).json({ msg: 'Cannot retrieve Marketplace Sale.' });
+    }
+}
+
+// Transaction between players
+exports.marketTransaction = async (req: $Request, res: $Response) => {
+    // Validation
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+        // Return 400 for a bad request
+        return res
+            .status(400)
+            .json({ errors: errors.array() });
+    }
+
+    // Deconstruct request body
+    const { buyerId, sellerId, saleId } = req.body;
+
+    // Check if users exist
+    let buyer = await User.findById(buyerId);
+    const seller = await User.findById(sellerId);
+    if (!buyer) {
+        return res  
+            .status(400)
+            .json({ msg: 'Buyer user does not exist' })
+    }
+
+    if (!seller) {
+        return res  
+            .status(400)
+            .json({ msg: 'Seller user does not exist' })
+    }
+
+    // Check if sale exists
+    const sale = await MarketSale.findById(saleId);
+    if (!sale) {
+        return res
+            .status(400)
+            .json({ msg: 'Sale post does not exist' })
+    }
+
+    // Check if buyer has enough money
+    if (buyer.money < sale.salePrice) {
+        return res
+            .status(400)
+            .json({ msg: 'Not enough money for this transaction.' })
+    }
+
+    // Buying a Tank or a Component
+    // Note: Depending on the itemType, if it is a tank,
+    // itemID is the ObjectId of the tank, otherwise it is
+    // the string identifier of the component/Casus block equivalent
+    // to the property in the User Model.
+    if (sale.itemType === 'tank') {
+        try {         
+            // Check if tank exists
+            const tank = await Tank.findById(sale.itemId);
+            if (!tank) {
+                return res
+                    .status(400)
+                    .json({ msg: 'Tank does not exist' })
+            }
+
+            // Start transaction
+            buyer = await User.findByIdAndUpdate(buyerId, { $inc: { money: (sale.salePrice * -1) } }, { new: true });
+            await User.findByIdAndUpdate(sellerId, { $inc: { money: sale.salePrice } });
+            await Tank.findByIdAndUpdate(sale.itemId, { $set: { userId: buyerId } });
+
+            // Remove the sale from database
+            await MarketSale.deleteOne({ id: saleId });
+
+            // Return current buyer
+            res.status(201).json(buyer);
+
+        } catch (err) {
+            console.error(err.message);
+            res.status(500).json({ msg: 'Cannot make Market Transaction.' });
+        }
+    } else { 
+        // Component 
+        if (sale.itemType === 'component') {       
+            try {
+                // Start transaction
+                buyer = await User.findByIdAndUpdate(buyerId, { $inc: { money: (sale.salePrice * -1) } }, { new: true });
```

Comment:
Just curious:
what does `new: true` do?

Reply:
`new: true` ensures the assigned object is now is the updated object. Or else after the change is made, the object (buyer) returned will have the properties before it was changed in the DB.